container.go

```diff
@@ -50,6 +53,20 @@ type Container struct {
 	graph.Graph
 }
 
+// InvokeOnce only allows function invocation once to register the
+// return types. If return types are already registered, specific error
+// is returned to the caller
+func (c *Container) InvokeOnce(t interface{}) error {
+	ctype := reflect.TypeOf(t)
+	if ctype.Kind() != reflect.Func {
+		return errors.Wrapf(errParamType, _invokeErr, ctype)
+	}
+	if err := c.ValidateReturnTypes(ctype, true); err != nil {
+		return ErrInvokeOnce
+	}
+	return c.Invoke(t)
```

Comment:
What do you mean by being recursive? There are no recursive calls with Invoke.
InvokeOnce is Invoking the function once, which is using `node.value()` to resolve dependencies recursively. InvokeOnce is just a wrapper around `Invoke` to make sure that object is not resolved multiple times. 

InvokeOnce returns `ErrInvokeOnce` which is used as a decider to skip double invokes.

fx.Start will be fixed with simple check - 
```go
if err := s.container.InvokeOnce(c); err != nil && err != dig.ErrInvokeOnce {
	return err
}
```